lnwallet/btcwallet/btcwallet.go

```diff
@@ -1199,33 +1200,98 @@ func (b *BtcWallet) ListUnspentWitness(minConfs, maxConfs int32,
+			alreadyConfirmedErr *base.ErrAlreadyConfirmed
+		)
+
 		// If we failed to publish the transaction, check whether we
 		// got an error of known type.
-		switch err.(type) {
+		switch {
 		// If the wallet reports a double spend, convert it to our
 		// internal ErrDoubleSpend and return.
-		case *base.ErrDoubleSpend:
+		case errors.As(err, &doubleSpendErr):
 			return lnwallet.ErrDoubleSpend
 
 		// If the wallet reports a replacement error, return
 		// ErrDoubleSpend, as we currently are never attempting to
 		// replace transactions.
-		case *base.ErrReplacement:
+		case errors.As(err, &replacementErr):
 			return lnwallet.ErrDoubleSpend
 
-		// If the wallet reports that fee requirements for accepting the
-		// tx into mempool are not met, convert it to our internal
+		// If the wallet reports that fee requirements for accepting
+		// the tx into mempool are not met, convert it to our internal
 		// ErrMempoolFee and return.
-		case *base.ErrMempoolFee:
+		case errors.As(err, &mempoolFeeErr):
 			return fmt.Errorf("%w: %v", lnwallet.ErrMempoolFee,
 				err.Error())
 
+		//nolint:lll
+		// These two errors are ignored inside `PublishTransaction`:
+		// https://github.com/btcsuite/btcwallet/blob/master/wallet/wallet.go#L3763
+		// To keep our current behavior, we need to ignore the same
+		// errors returned from TestMempoolAccept.
+		//
+		// TODO(yy): since `LightningWallet.PublishTransaction` always
+		// publish the same tx twice, we'd always get ErrInMempool. We
+		// should instead create a new rebroadcaster that monitors the
+		// mempool, and only rebroadcast when the tx is evicted. This
+		// way we don't need to broadcast twice, and can instead return
+		// these errors here.
+		case errors.As(err, &inMempoolErr),
+			errors.As(err, &alreadyConfirmedErr):
+
+			return nil
+
 		default:
 			return err
 		}
 	}
-	return nil
+
+	// For neutrino backend there's no mempool, so we return early by
+	// publishing the transaction.
+	if b.chain.BackEnd() == "neutrino" {
+		err := b.wallet.PublishTransaction(tx, label)
+		return handleErr(err)
+	}
+
+	// For non-neutrino nodes, we will first check whether the transaction
+	// can be accepted by the mempool.
+	// Use a max feerate of 0 means the default value will be used when
+	// testing mempool acceptance. The default max feerate is 0.10 BTC/kvb,
+	// or 10,000 sat/vb.
+	results, err := b.chain.TestMempoolAccept([]*wire.MsgTx{tx}, 0)
+	if err != nil {
+		return err
+	}
+
+	result := results[0]
+	log.Debugf("TestMempoolAccept result: %s", spew.Sdump(result))
+
+	// If the check failed, there's no need to publish it. We'll
+	// handle the error and return.
+	if !result.Allowed {
+		log.Warnf("Transaction %v not accepted by mempool: %v",
+			tx.TxHash(), result.RejectReason)
+
+		// We need to use the string to create an error type
+		// and map it to a btcwallet error.
+		err := base.MapBroadcastBackendError(
+			errors.New(result.RejectReason),
+		)
+
+		return handleErr(err)
```

Comment:
Now that we can detect failed publish attempts, do we need to handle this at the sweeper level? Let's say the sweeper tries to publish something, but it violates an RBF rule, should it try to bump the fee until accepted? 

Also we should consider the impact on normal callers here that happen to be using the API for their own higher level protocols. With this change, they'll now get a new error where otherwise things would succeed. This may lead to us exposing richer errors for publish failures on the RPC level as well.

Reply:
Yeah I wanna implement this https://github.com/btcsuite/btcd/issues/2101 so we have a richer error mapping so we can have a finer control over the publishing process here and in `btcwallet` when publishing tx. The sweeper will call `testmempoolaccept` before publishing txns to handle RBF, and it won't be handled here as RBF cannot be performed in some of the cases where `PublishTransaction` are called, like FC or chan opening. 

As for here, the main usage of `testmempoolaccept` is to make sure we can bail early before even hitting other logics inside `btcwallet`, such as saving to or updating the db. 

> Also we should consider the impact on normal callers here that happen to be using the API for their own higher level protocols.

Yeah indeed. The errors returned from `testmempoolaccept` are matched using the same matching logic applied in `btcwallet.PublishTransaction` to minimize the disturbance. But again, we need to have richer errors here and handle them. As for now, I think it should behave the same for most of the cases, such as double spending errors, input missing errors, below mempool fee errors, etc.